Encapsulation support for standalone loadbalancer

mchalla opened the pull request:
- Standalone lb initialized with tunneled or direct mode.
- Mode change requires restarting standalone lb.
- The mode is decided based on ENCAP_IFINDEX that is generated by
  daemon if vxlan is used
- Added LB encap tests

Fixes: #282

jrfastab commented on bpf/bpf_lb.c:
```diff
@@ -102,18 +121,40 @@ static inline int handle_ipv6(struct __sk_buff *skb)
-	return TC_ACT_REDIRECT;
+#ifdef ENCAP_IFINDEX
+	ret = lb6_encap(skb, &new_dst, svc->rev_nat_index);
+	if (IS_ERR(ret))
```
The use of IS_ERR(ret) here seems a bit odd to me just return the value is checked in `from_netdev`. Besides a few lines lower its returned. For what its worth IS_ERR is overly used throughout so copying the pattern is understandable.

tgraf replied:
I had `goto drop` before I rebased with @mchalla's changes and that change got lost.

tgraf replied:
Fixed

jrfastab commented on bpf/bpf_lb.c:
```diff
@@ -164,7 +205,15 @@ static inline int handle_ipv4(struct __sk_buff *skb)
-	return TC_ACT_REDIRECT;
+#ifdef ENCAP_IFINDEX
+	ret = lb4_encap(skb, new_dst, svc->rev_nat_index);
+	if (IS_ERR(ret))
```
Same here....

jrfastab commented on bpf/bpf_lb.c:
```diff
@@ -194,9 +243,14 @@ int from_netdev(struct __sk_buff *skb)
+#else /* ENCAP_IFINDEX */
 
-#ifdef LB_REDIRECT
+#if defined LB_REDIRECT
```
why the churn from `ifdef` to `if defined` do you plan for LB_REDIRECT to be compound conditional

tgraf replied:
This is a leftover, I'll drop this change.

tgraf replied:
Fixed

jrfastab commented on tests/06-lb-encap.sh:
```diff
@@ -0,0 +1,192 @@
+cat <<EOF | cilium -D policy import -
+[{
+    "endpointSelector": {"matchLabels":{"id.server":""}},
+    "ingress": [{
```
General comment (not needed for this PR)  we should be adding ingress and egress policies to all tests so we catch both code paths.

tgraf replied:
Agreed. The only egress policy we could add here is CIDR. I worry a bit that if this test fails because of a CIDR bug, we won't notice that and would mistake it for an encap specific bug.

jrfastab reviewed (changes requested):
Couple small style comments otherwise looks good.

tgraf commented on bpf/node_config.h:
```diff
@@ -43,3 +42,4 @@
 #define NODE_MAC { .addr = { 0xde, 0xad, 0xbe, 0xef, 0xc0, 0xde } }
 #define ENABLE_IPV4
 #define LB_RR_MAX_SEQ 31
+#define GENEVE_OPTIONS_WORLD { 0xff, 0xff, 0x1, 0x1, 0x0, 0x0, 0x0, 0x2 }
```
Noticed this left over as well now, will remove.

tgraf committed:
bpf: Update bindata

Signed-off-by: Thomas Graf <thomas@cilium.io>

mchalla committed:
bpf: Encapsulation support for standalone loadbalancer

- Standalone lb initialized with tunneled or direct mode.
- Mode change requires restarting standalone lb.
- The mode is decided based on ENCAP_IFINDEX that is generated by
  daemon if vxlan is used
- Added LB encap tests

Signed-off-by: Madhu Challa <madhu@cilium.io>
Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Disable rp_filter on cilium_host

When doing DSR back to an IP address which is configured locally, we may
be seeing packets on the cilium_host address from a source address to
which the route does not point to the cilium_host device itself.

Signed-off-by: Thomas Graf <thomas@cilium.io>

jrfastab reviewed (approved):
Looks good.
